src/ast/semantic_analyser.cpp

```diff
@@ -462,6 +462,50 @@ void SemanticAnalyser::visit(Call &call)
       }
     }
   }
+  else if (call.func == "buf")
+  {
+    if (!check_varargs(call, 1, 2))
+      return;
+
+    auto &arg = *call.vargs->at(0);
+    if (!(arg.type.type == Type::integer || arg.type.type == Type::string ||
+          arg.type.type == Type::array))
+      error(call.func +
+                "() expects an integer, string, or array argument but saw " +
+                typestr(arg.type.type),
+            call.loc);
+
+    size_t max_buffer_size = bpftrace_.strlen_;
+    size_t buffer_size = max_buffer_size;
+
+    if (call.vargs->size() == 1)
+      if (arg.type.type == Type::array)
+        buffer_size = arg.type.pointee_size * arg.type.size;
+      else
+        error(call.func + "() expects a length argument for non-array type " +
+                  typestr(arg.type.type),
+              call.loc);
+    else
+    {
+      if (is_final_pass())
+        check_arg(call, Type::integer, 1, false);
+
+      auto &size_arg = *call.vargs->at(1);
+      if (size_arg.is_literal)
+        buffer_size = static_cast<Integer &>(size_arg).n;
+    }
+
+    if (buffer_size > max_buffer_size)
+      buffer_size = max_buffer_size;
+
+    buffer_size++; // allow room for zero byte
```

Comment:
Why do we need a zero byte? We already have its length?

Reply:
I misunderstood what was happening here when I wrote the comment. I think we need the extra byte because otherwise `arg.type.size` will be one byte too small (because I'm using a byte to embed the buffer length), and the final print buffer will be corrupted:

https://github.com/iovisor/bpftrace/blob/aa94d9b363b70f9ed0f43d2345605bf7880f099c/src/ast/codegen_llvm.cpp#L1973